Feature/strings

0xdea opened the pull request:
I've implemented the [strlist](https://cpp.docs.hex-rays.com/strlist_8hpp.html) API (and a conversion function that I needed while I was at it). Let me know if I need to modify some things before it can be merged. Thanks!

xorpse commented on idalib/src/idb.rs:
```diff
@@ -379,6 +381,29 @@ impl IDB {
+        StringList::new(self)
+    }
+
+    pub fn get_string(&self, index: usize) -> Option<String> {
```
I'd move this to `StringList` as `get_by_index`

0xdea replied:
Sure, the only problem I have is that I need to access the `get_bytes()` method from within `StringList`. Is there a clean way to achieve this with the current architecture that uses `PhantomData`?

xorpse replied:
I guess we have two options:
- Reuse the implementation of `get_bytes` inline inside this function. If we find this is a repeated pattern later, we can figure out how to reuse it.
- Implement the logic in C++ such that we return a String for a given index.

0xdea replied:
Done! I went for the first option.

xorpse commented on idalib/src/idb.rs:
```diff
@@ -379,6 +381,29 @@ impl IDB {
+
+        let bytes = self.get_bytes(addr, len);
+
+        Some(String::from_utf8_lossy(&bytes).escape_debug().to_string())
```
This could be `String::from_utf8_lossy(&bytes).into_owned()`

0xdea replied:
Ok. It is my understanding that we still have to use `to_string()` after `escape_debug()`. Or am I missing something?

xorpse replied:
I don’t think you need `escape_debug` at all here, if the intention is to just get a valid utf8 string.

0xdea replied:
I see. It makes sense. I applied the escape because I wanted to be able to print strings in a similar way as IDA does in its GUI. But users of idalib can always decide to escape or not on their own. I'll modify it as soon as I get to a computer (I'm currently on mobile).

xorpse commented on idalib/src/idb.rs:
```diff
@@ -379,6 +381,29 @@ impl IDB {
+        Some(String::from_utf8_lossy(&bytes).escape_debug().to_string())
+    }
+
+    pub fn ea2str(&self, ea: Address) -> Option<String> {
```
I'd move this to `StringList` as `get_by_address`

0xdea replied:
This function actually just converts an address into its rich string representation. For instance:
```
ea2str(0x9141) = ".rodata:0000000000009141"
```
Its name in the IDA API is somewhat misleading and that's why I implemented it in the first place. Then I noticed it doesn't do what I thought it did and it doesn't have much to do with strings in the `StringList`, so I moved it to a `conversions` module that we can populate in the future with other similar utility functions. We could also remove it to keep idalib cleaner if you prefer.

xorpse replied:
Hmm maybe we can call it something less misleading on our side—I’ve got no issue keeping it in that case :)

0xdea replied:
Done. I'll implement other conversion utility functions that make sense when I have time (e.g., some of those from <https://cpp.docs.hex-rays.com/group__conv.html>).

xorpse commented on idalib/src/strings.rs:
```diff
@@ -0,0 +1,54 @@
+    }
+
+    // Note: Seems pretty much useless
+    pub fn build(&self) {
```
If these aren't needed, then maybe we can just not expose them?

0xdea replied:
I could remove them if you want. From my tests they don't do anything, but IDA SDK's documentation [here](https://cpp.docs.hex-rays.com/strlist_8hpp.html) seems to imply that they're useful, so I kept them just in case I had missed something during testing:
```
While the kernel keeps the string list, it does not update it. The string list is 
not used by the kernel because keeping it up-to-date would slow down IDA 
without any benefit. If the string list is not cleared using clear_strlist(), the 
list will be saved to the database and restored on the next startup.

The users of this list should call build_strlist() if they need an up-to-date 
version.
```
Let me know if I should remove them for the time being.

xorpse replied:
Maybe we can expose `build` as `rebuild` since that's what this description seems to imply is going on? And keep both it and `clear`?

0xdea replied:
Done!

xorpse commented on idalib/src/strings.rs:
```diff
@@ -0,0 +1,54 @@
+        unsafe { clear_strlist() }
+    }
+
+    pub fn get_item_addr(&self, index: usize) -> Option<Address> {
```
I guess this is and `get_item_length` are low-level primitives so we can just remove the `pub` visibility marker and keep them internal.

0xdea replied:
I think our API should provide a way to get the address of each string in the `StringList`. Also `get_item_length()` might have its use. Not sure about the type that's also shown in the Strings windows of IDA's UI, that's why I haven't implemented it.

I think it would make sense to keep `get_item_addr()` public with a clearer name (do you have any recommedations?) and make `get_item_length()` private. I could also implement a private `get_item_type()` for future use (types are integers that should be interpreted and I didn't look into it yet).

As an alternative, I could modify `get_by_index()` to return an `(Address, String)` tuple. Please advise on your preference.

xorpse replied:
Let’s keep `get_item_addr` as `get_address_by_index`

xorpse commented on idalib/src/strings.rs:
```diff
@@ -0,0 +1,54 @@
+pub struct StringList<'a> {
+    _marker: PhantomData<&'a IDB>,
+}
+
```
I'd add a type alias for `StringIndex` as we do for bookmarks to keep consistency.

0xdea replied:
Done!

xorpse reviewed (changes requested):
Thanks once again for the PR :)! I've left some comments to make the API a bit more consistent/idiomatic, nothing major.

0xdea commented:
Thank you so much for the review! Please see my comments. Cheers!

xorpse commented on idalib/src/idb.rs:
```diff
@@ -379,6 +381,31 @@ impl IDB {
+        Some(String::from_utf8_lossy(&bytes).into_owned())
+    }
+
+    pub fn convert_ea2str(&self, ea: Address) -> Option<String> {
```
Maybe let's call it `address_to_string`?

0xdea replied:
Yeah, much better. Done!

0xdea commented:
Thanks for the review. I've moved `IDB::get_string` to `StringList::get_by_index`, exposed `StringList::build` as `StringList::rebuild`, removed `pub` from `get_length_by_index`, and renamed `convert_ea2str` to `address_to_string`. We should be good to go!

xorpse reviewed (approved):
Thank you! All looks good now :)
